src/server/pps/server/api_server.go

```diff
@@ -1875,22 +1913,135 @@ func (a *apiServer) CreatePipelineInTransaction(txnCtx *txnenv.TransactionContex
+	pipelinePtr := pps.StoredPipelineInfo{}
+	if err := a.pipelines.ReadWrite(txnCtx.SqlTx).Get(pipelineName, &pipelinePtr); err != nil {
+		if col.IsErrNotFound(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+	// the spec commit must already exist outside of the transaction, so we can retrieve it normally
+	return ppsutil.GetPipelineInfo(txnCtx.Client, &pipelinePtr)
+}
+
+func (a *apiServer) pipelineInfosForUpdate(txnCtx *txnenv.TransactionContext, request *pps.CreatePipelineRequest) (*pps.PipelineInfo, *pps.PipelineInfo, error) {
+	if request.Pipeline == nil {
+		return nil, nil, errors.New("invalid pipeline spec: request.Pipeline cannot be nil")
+	}
+
+	// We can't recover from an incomplete pipeline info here because
+	// modifying the spec repo depends on being able to access the previous
+	// commit. We therefore use `GetPipelineInfo` which will error if the
+	// spec commit isn't working.
+	oldPipelineInfo, err := a.latestPipelineInfo(txnCtx, request.Pipeline.Name)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	newPipelineInfo, err := a.initializePipelineInfo(request, oldPipelineInfo)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return oldPipelineInfo, newPipelineInfo, nil
+}
+
+func (a *apiServer) CreatePipelineInTransaction(
+	txnCtx *txnenv.TransactionContext,
+	request *pps.CreatePipelineRequest,
+	specFilesetID *string,
+	prevSpecCommit **pfs.Commit,
+) error {
+	oldPipelineInfo, newPipelineInfo, err := a.pipelineInfosForUpdate(txnCtx, request)
+	if err != nil {
 		return err
 	}
+	pipelineName := request.Pipeline.Name
+
+	if *specFilesetID != "" {
+		// If we already have a fileset, try to renew it - if that fails, invalidate it
+		if err := txnCtx.Client.RenewFileSet(*specFilesetID, 600*time.Second); err != nil {
+			*specFilesetID = ""
+		}
+	}
+
+	// If the expected spec commit doesn't match up with oldPipelineInfo, we need
+	// to recreate the fileset
+	staleFileset := false
+	if oldPipelineInfo == nil {
+		staleFileset = (*prevSpecCommit != nil)
+	} else {
+		staleFileset = !proto.Equal(oldPipelineInfo.SpecCommit, *prevSpecCommit)
+	}
+
+	if staleFileset || *specFilesetID == "" {
+		// No existing fileset or the old one expired, create a new fileset - the
+		// pipeline spec to be written into a fileset outside of the transaction.
+		*specFilesetID, err = a.writePipelineInfoToFileset(txnCtx.Client, newPipelineInfo)
+		if err != nil {
+			return err
+		}
+		if oldPipelineInfo != nil {
+			*prevSpecCommit = oldPipelineInfo.SpecCommit
+		}
 
+		// The transaction cannot continue because it cannot see the fileset - abort and retry
+		return &col.ErrTransactionConflict{}
+	}
+
+	// Verify that all input repos exist (create cron and git repos if necessary)
 	var visitErr error
-	pps.VisitInput(pipelineInfo.Input, func(input *pps.Input) {
+	pps.VisitInput(newPipelineInfo.Input, func(input *pps.Input) {
+		if input.Pfs != nil {
+			if _, err := txnCtx.Pfs().InspectRepoInTransaction(txnCtx,
```

Comment:
Moved out of `validatePipelineInTransaction` so that it no longer needed to make service calls (and we're already verifying the existence of inputs here).